homeassistant/components/evohome/__init__.py

```diff
@@ -44,204 +45,243 @@
+        else:
+            if access_token != self.client.access_token:
+                await self._save_auth_tokens()
+
+        finally:
+            self.params[CONF_PASSWORD] = 'REDACTED'
 
-    def _handle_exception(self, err):
+        loc_idx = self.params[CONF_LOCATION_IDX]
         try:
-            raise err
+            self.config = client.installation_info[loc_idx][GWS][0][TCS][0]
 
-        except evohomeclient2.AuthenticationError:
+        except IndexError:
             _LOGGER.error(
-                "Failed to (re)authenticate with the vendor's server. "
-                "This may be a temporary error. Message is: %s",
-                err
+                "Config error: '%s' = %s, but its valid range is 0-%s. "
+                "Unable to continue. "
+                "Fix any configuration errors and restart HA.",
+                CONF_LOCATION_IDX, loc_idx, len(client.installation_info) - 1
             )
+            return False
 
-        except requests.exceptions.ConnectionError:
-            # this appears to be common with Honeywell's servers
-            _LOGGER.warning(
-                "Unable to connect with the vendor's server. "
-                "Check your network and the vendor's status page."
-            )
+        else:
+            self.tcs = \
+                client.locations[loc_idx]._gateways[0]._control_systems[0]  # noqa: E501; pylint: disable=protected-access
+
+        _LOGGER.debug("Config = %s", self.config)
 
-        except requests.exceptions.HTTPError:
-            if err.response.status_code == HTTP_SERVICE_UNAVAILABLE:
-                _LOGGER.warning(
-                    "Vendor says their server is currently unavailable. "
-                    "This may be temporary; check the vendor's status page."
-                )
-
-            elif err.response.status_code == HTTP_TOO_MANY_REQUESTS:
-                _LOGGER.warning(
-                    "The vendor's API rate limit has been exceeded. "
-                    "So will cease polling, and will resume after %s seconds.",
-                    (self._params[CONF_SCAN_INTERVAL] * 3).total_seconds()
-                )
-                self._timers['statusUpdated'] = datetime.now() + \
-                    self._params[CONF_SCAN_INTERVAL] * 3
+        return True
 
+    async def _load_auth_tokens(self) -> Tuple[str, str, datetime]:
+        store = self.hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+        app_storage = self._app_storage = await store.async_load()
+
+        if app_storage.get(CONF_USERNAME) == self.params[CONF_USERNAME]:
+            refresh_token = app_storage.get(CONF_REFRESH_TOKEN)
+            access_token = app_storage.get(CONF_ACCESS_TOKEN)
+            at_expires_str = app_storage.get(CONF_ACCESS_TOKEN_EXPIRES)
+            if at_expires_str:
+                at_expires_dt = as_utc(parse_datetime(at_expires_str))
+                at_expires_dt = at_expires_dt.astimezone(tzlocal())
```

Comment:
I tried to do this without using `dateutil.tztzlocal()` and instead using `util.dt.get_time_zone()`, but couldn't get it to work.

I am calling a web API and want to convert a local time (`access_token_expires`) into UTC (to be stored/restored, and) used for `async_track_point_in_utc_time()`

Only one other component uses `get_time_zone`, but gets the string from **configuration.yaml**, which is not satisfactory:
```python
./components/worldclock/sensor.py:30:    
time_zone = dt_util.get_time_zone(config.get(CONF_TIME_ZONE))
```
... as I need the offset used by the evohome API, and not the TZ as specified by the user.

Reply:
Is there any reason why `access_token_expires` is in local?

Reply:
Because local is the TZ used by the client API and the dt is naive.  The client API uses `access_token_expires` to know when to re-authorize, and the integration uses it to know when it is time to store the new tokens.